plugins/outputs/whatap/whatap.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/outputs"
+
+	wio "github.com/whatap/go-api/common/io"
+	wpack "github.com/whatap/go-api/common/lang/pack"
+	whash "github.com/whatap/go-api/common/util/hash"
+)
+
+const (
+	NetSrcAgentOneway  = 10
+	NetSrcAgentVersion = 0
+	Scheme             = "tcp"
+)
+
+type Whatap struct {
+	License string          `toml:"license"`
+	Servers []string        `toml:"servers"`
+	Pcode   int64           `toml:"project_code"`
+	Timeout config.Duration `toml:"timeout"`
+	Log     telegraf.Logger `toml:"-"`
+	oname   string
+	oid     int32
+	conn    net.Conn
+	dest    int
+	hosts   []string
+}
+
+const sampleConfig = `
+  ## You can create a project on the WhaTap site(https://www.whatap.io) 
+  ## to get license, project code and server IP information.
+
+  ## WhaTap license. Required
+  license = "xxxx-xxxx-xxxx"
+
+  ## WhaTap project code. Required
+  project_code = 1111
+
+  ## WhaTap server IP. Required
+  ## Put multiple IPs. ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+  servers = ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+
+  ## Connection timeout.
+  # timeout = "60s"
+`
+
+func (w *Whatap) Connect() error {
+	// Change and connect multiple servers sequentially.
+	host := w.nextHost()
+	client, err := net.DialTimeout(Scheme, host, time.Duration(w.Timeout))
+	if err != nil {
+		return err
+	}
+	w.conn = client.(*net.TCPConn)
+	w.Log.Info("Connected ", host)
+	return nil
+}
+
+func (w *Whatap) Close() error {
+	if w.conn == nil {
+		return nil
+	}
+	err := w.conn.Close()
+	w.conn = nil
+	return err
+}
+
+func (w *Whatap) Description() string {
+	return "Plugin to send metrics to a WhaTap server"
+}
+
+func (w *Whatap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (w *Whatap) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+	if w.conn == nil {
+		if err := w.Connect(); err != nil {
+			return err
+		}
+	}
+	// Transform telegraf metrics to whatap data.
+	for _, m := range metrics {
+		p := wpack.NewTagCountPack()
+		p.Pcode = w.Pcode
+		p.Oid = w.oid
+		p.Category = "telegraf_" + m.Name()
+		for k, v := range m.Fields() {
+			p.Put(k, v)
+		}
+		for k, v := range m.Tags() {
+			p.PutTag(k, v)
+		}
+		p.PutTag("oname", w.oname)
+
+		// Convert time to microseconds.
+		p.Time = m.Time().UnixNano() / int64(time.Millisecond)
+
+		dout := wio.NewDataOutputX()
+		dout.WriteShort(p.GetPackType())
+		p.Write(dout)
+
+		if err := w.send(dout.ToByteArray()); err != nil {
+			w.Log.Warnf("cannot send data: %v", err)
+			_ = w.Close()
+		}
+	}
+	return nil
+}
+func (w *Whatap) send(b []byte) (err error) {
+	dout := wio.NewDataOutputX()
+	// Header : add the header before sending
+	dout.WriteByte(NetSrcAgentOneway)
+	dout.WriteByte(NetSrcAgentVersion)
+	dout.WriteLong(w.Pcode)
+	dout.WriteLong(whash.Hash64Str(w.License))
+	// Body : Set the converted whatap data(b []bytes) as the body
+	dout.WriteIntBytes(b)
+	sendbuf := dout.ToByteArray()
```

Comment:
In line 111-113 you are creating a `dout` only containing `p` as far as I do understand the code, then in `send` you create a new `dout` where you add a header and the the `dout` you created before. Why don't you create the **complete** `dout` (i.e. the second one) _before passing it into `send`? This would have been more intuitive and would group all `Write...` functions in one place...

Reply:
Edited and applied. We ask for confirmation. I modified the WriteHeader function by creating it in an external library.